llvm/lib/Transforms/ObjCARC/ObjCARCOpts.cpp

```diff
@@ -2485,6 +2490,203 @@ bool ObjCARCOpt::run(Function &F, AAResults &AA) {
+    return false;
+
+  for (User *U : Val->users()) {
+    if (U == Target) {
+      return true;
+    }
+
+    // For pointer casts, recursively check their users
+    if (isa<CastInst>(U) || isa<BitCastInst>(U)) {
+      Value *CastResult = dyn_cast<Value>(U);
+      if (CastResult &&
+          checkLeadsToTarget(CastResult, Target, MaxDepth - 1, Visited)) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
+
+/// Optimize autorelease pools by eliminating empty push/pop pairs.
+void ObjCARCOpt::OptimizeAutoreleasePools(Function &F) {
+  LLVM_DEBUG(dbgs() << "\n== ObjCARCOpt::OptimizeAutoreleasePools ==\n");
+
+  OptimizationRemarkEmitter ORE(&F);
+
+  // Track empty autorelease pool push/pop pairs
+  SmallVector<std::pair<CallInst *, CallInst *>, 4> EmptyPoolPairs;
+
+  // Process each basic block independently.
+  // TODO: Can we optimize inter-block autorelease pool pairs?
+  // This would involve tracking autorelease pool state across blocks.
+  for (BasicBlock &BB : F) {
+    // Use a stack to track nested autorelease pools
+    SmallVector<std::pair<CallInst *, bool>, 4>
+        PoolStack; // {push_inst, has_autorelease_in_scope}
+
+    for (Instruction &Inst : BB) {
+      ARCInstKind Class = GetBasicARCInstKind(&Inst);
+
+      switch (Class) {
+      case ARCInstKind::AutoreleasepoolPush: {
+        // Start tracking a new autorelease pool scope
+        CallInst *Push = cast<CallInst>(&Inst);
+        PoolStack.push_back(
+            {Push, false}); // {push_inst, has_autorelease_in_scope}
+        LLVM_DEBUG(dbgs() << "Found autorelease pool push: " << *Push << "\n");
+        break;
+      }
+
+      case ARCInstKind::AutoreleasepoolPop: {
+        CallInst *Pop = cast<CallInst>(&Inst);
+
+        if (PoolStack.empty())
+          break;
+
+        auto &TopPool = PoolStack.back();
+        CallInst *PendingPush = TopPool.first;
+        bool HasAutoreleaseInScope = TopPool.second;
+
+        // Check if this pop matches the pending push by comparing the token
+        Value *PopArg = Pop->getArgOperand(0);
+        bool IsMatchingPop = (PopArg == PendingPush);
+
+        // Also handle pointer casts by stripping them
+        if (!IsMatchingPop) {
+          Value *StrippedPopArg = PopArg->stripPointerCasts();
+          IsMatchingPop = (StrippedPopArg == PendingPush);
+        }
+
+        // Before adding to EmptyPoolPairs, verify ALL uses of the push
+        // eventually lead to this specific pop instruction
+        bool SafeToOptimize = true;
+        if (IsMatchingPop) {
+          for (User *U : PendingPush->users()) {
+            // Check if this use eventually leads to our pop instruction
+            bool LeadsToExpectedPop = false;
+
+            if (U == Pop) {
+              LeadsToExpectedPop = true;
+            } else {
+              // For pointer casts, check if they eventually lead to the pop
+              // (up to 3 levels deep)
+              Value *CastResult = dyn_cast<Value>(U);
```

Comment:
ditto on this `dyn_cast`: User's are Value's, so this can't fail. maybe you want `auto *Cast = dyn_cast<CastInst>(U)` instead?

Reply:
that can be combined into the preceding `else` as an `else if (`